pkg/deploy.go

```diff
@@ -20,71 +20,43 @@ type deployWorkflow struct {
 
 func (w *deployWorkflow) start(ctx context.Context, action <-chan event, release, trigger chan<- event) error {
 	log := w.log.WithName("deploy")
-OUTER:
 	for {
 		select {
 		case action := <-action:
 			log.Info("deploy started...")
 			switch action.id {
 			case triggeredMessage:
-				vars := []envVar{}
-				vars = append(vars, action.envs...)
-				m, err := groupEnvs(vars...)
-				if err != nil {
+				vars := newEnvVars(action.envs...)
+				if _, err := vars.toMap(); err != nil {
 					log.Error(err, "could not map envs")
 					trigger <- event{id: deployedMessage}
-					continue OUTER
+					continue
 				}
-				artifactDirectory, err := replaceEnvs(w.Artifact.Directory, m)
+				artifactDirectory, err := vars.replace(w.Artifact.Directory)
 				if err != nil {
 					log.Error(err, "could not transform directory")
 					trigger <- event{id: deployedMessage}
-					continue OUTER
+					continue
 				}
 				artifactDirectory = path.Join(w.execdir, artifactDirectory)
-				err = os.MkdirAll(artifactDirectory, os.ModeDir|os.ModePerm)
-				if err != nil {
+				if err := os.MkdirAll(artifactDirectory, os.ModeDir|os.ModePerm); err != nil {
 					log.Error(err, "could not create directory", "data", artifactDirectory)
 					trigger <- event{id: deployedMessage}
-					continue OUTER
+					continue
 				}
-				vars = append(vars, envVar{Name: "artifactDirectory", Value: artifactDirectory})
-				m["artifactDirectory"] = artifactDirectory
-				artifactFilename, err := replaceEnvs(w.Artifact.Filename+w.Artifact.Extension, m)
+				vars.addOne("artifactDirectory", artifactDirectory)
+				artifactFilename, err := vars.replace(w.Artifact.Filename + w.Artifact.Extension)
 				if err != nil {
 					log.Error(err, "could not transform filename")
 					trigger <- event{id: deployedMessage}
-					continue OUTER
+					continue
 				}
-				vars = append(vars, envVar{Name: "artifactFilename", Value: artifactFilename})
-				m["artifactFilename"] = artifactFilename
+				vars.addOne("artifactFilename", artifactFilename)
 				artifact := path.Join(artifactDirectory, artifactFilename)
-				vars = append(vars, envVar{Name: "artifact", Value: artifact})
-				m["artifact"] = artifact
-				for _, v := range w.flow {
-					cmd := w.keys[v]
-					cmd.log = log
-					if cmd.Command == "" {
-						continue
-					}
-					log.Info("running...", "data", v)
-					e, err := (&cmd).run(w.workspace, m)
-					if err != nil {
-						w.state.notifyStep(
-							getEnv(action.envs, "version"), "deploy",
-							runnerStatusFailed,
-							step{execStruct: cmd, Name: v})
-						log.Error(err, "execution error")
-						trigger <- event{id: deployedMessage}
-						continue OUTER
-					}
-					w.state.notifyStep(
-						getEnv(action.envs, "version"), "deploy",
-						runnerStatusDone,
-						step{execStruct: cmd, Name: v})
-					if e != nil {
-						vars = append(vars, *e)
-					}
+				vars.addOne("artifact", artifact)
+
+				if err := w.startFlows(action, &vars); err != nil {
+					continue
```

Comment:
Not sure where to put it but there is a need to keep `trigger <- event{id: deployedMessage}`. The basic idea is that the `trigger` workflow stacks changes and only push them when the `deploy` is available. As a result, the `trigger` flow keep track of the fact the `deploy` is running or not. The counterpart of that is that the `deploy` must notify the trigger of the fact is has finished the work. The fact that it has failed/succeeded is irrelevant for the `trigger`.